server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2296,13 +2283,19 @@ public void acquireReplicaOperationPermit(final long opPrimaryTerm, final long g
                                 } else {
                                     localCheckpoint = currentGlobalCheckpoint;
                                 }
-                                logger.trace(
-                                    "detected new primary with primary term [{}], resetting local checkpoint from [{}] to [{}]",
-                                    opPrimaryTerm,
-                                    getLocalCheckpoint(),
-                                    localCheckpoint);
-                                getEngine().resetLocalCheckpoint(localCheckpoint);
-                                getEngine().rollTranslogGeneration();
+                                if (isEngineResetting()) {
+                                    engineHolder.closePendingEngine();
```

Comment:
I think this is okay because we start another engine after that in this case. Moreover, we stick the "reset" local checkpoint (expose the local checkpoint of the active engine) while resetting the engine; thus the global checkpoint won't advance.